streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -79,11 +79,23 @@ public void register(final StateRestorer restorer) {
             return completed();
         }
 
-        final Set<TopicPartition> restoringPartitions = new HashSet<>(needsRestoring.keySet());
         try {
-            final ConsumerRecords<byte[], byte[]> allRecords = restoreConsumer.poll(10);
-            for (final TopicPartition partition : restoringPartitions) {
-                restorePartition(allRecords, partition, active.restoringTaskFor(partition));
+            final ConsumerRecords<byte[], byte[]> records = restoreConsumer.poll(10);
+            final Iterator<TopicPartition> iterator = needsRestoring.keySet().iterator();
+            final HashSet<TopicPartition> completedPartitions = new HashSet<>();
+            while (iterator.hasNext()) {
+                final TopicPartition partition = iterator.next();
+                final StateRestorer restorer = stateRestorers.get(partition);
+                final long pos = processNext(records.records(partition), restorer, updatedEndOffsets.get(partition));
+                restorer.setRestoredOffset(pos);
+                if (restorer.hasCompleted(pos, updatedEndOffsets.get(partition))) {
+                    restorer.restoreDone();
+                    updatedEndOffsets.remove(partition);
+                    completedPartitions.add(partition);
```

Comment:
One more thing -- missed this before: can't we remove `completedPartitions` and call `iterator.remove()` instead?

Maybe we need to change the iterator to iterate over the `HashMap` instead of the "key-set" of the HashMap though.